### PR TITLE
add Msg variants of the Request() functions for headers

### PR DESCRIPTION
**Pull request by ripienaar**
Signed-off-by: R.I.Pienaar <rip@devco.net>

### CODE REVIEW

**Comment by ripienaar** [comment]
In progress work to make `Request()` related functions support msg like `PublishMsg()`, probably dont need a ton of tests around this but will pick this up again tomorrow unless someone finish this before me

File: context.go [hunk 1]
```diff
@@ -63,11 +61,18 @@ func (nc *Conn) RequestWithContext(ctx context.Context, subj string, data []byte
...
-	return msg, nil
+	return rmsg, nil
 }
 
-// oldRequestWithContext utilizes inbox and subscription per request.
-func (nc *Conn) oldRequestWithContext(ctx context.Context, subj string, data []byte) (*Msg, error) {
+// RequestWithContext takes a context, a subject and payload
+// in bytes and request expecting a single response.
+func (nc *Conn) RequestWithContext(ctx context.Context, subj string, data []byte) (*Msg, error) {
+	msg := NewMsg(subj)
```
Comment (derekcollison) [inline context.go:70]: Would not penalize non-msg based one with extra msg IMO.

File: nats.go [hunk 2]
```diff
@@ -2873,8 +2873,7 @@ func (nc *Conn) respHandler(m *Msg) {
 	}
 }
 
-// Helper to setup and send new request style requests. Return the chan to receive the response.
-func (nc *Conn) createNewRequestAndSend(subj string, data []byte) (chan *Msg, string, error) {
+func (nc *Conn) createNewRequestAndSendMsg(msg *Msg) (chan *Msg, string, error) {
```
Comment (derekcollison) [inline nats.go:2876]: Would do hdr, msg []byte at lower levels and avoid msg allocation.

File: nats.go [hunk 3]
```diff
@@ -2898,16 +2897,16 @@ func (nc *Conn) createNewRequestAndSend(subj string, data []byte) (chan *Msg, st
...
 		return nil, token, err
 	}
 
 	return mch, token, nil
 }
 
-// Request will send a request payload and deliver the response message,
-// or an error, including a timeout if no message was received properly.
-func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg, error) {
+func (nc *Conn) RequestMsg(msg *Msg, timeout time.Duration) (*Msg, error) {
```
Comment (ripienaar) [inline nats.go:2909]: needs a comment here

File: nats.go [hunk 4]
```diff
@@ -2942,13 +2948,26 @@ func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg,
...
+}
+
+// Request will send a request payload and deliver the response message,
+// or an error, including a timeout if no message was received properly.
+func (nc *Conn) Request(subj string, data []byte, timeout time.Duration) (*Msg, error) {
+	if nc == nil {
+		return nil, ErrInvalidConnection
+	}
+
+	msg := NewMsg(subj)
```
Comment (derekcollison) [inline nats.go:2954]: Same here.

**Comment by ripienaar** [comment]
Redid this and also added a `RespondMsg()`, added tests

**Pull request merged**